Feature-187-refactor-severa-api-service

StudVladimir opened the pull request:
Refactor CreateSeveraApiService.

Put next functions to interface and put implementation inside CreateSeveraApiService.

getResourceAllocationsByUserOrAll()
checkKeywordExists()
fetchUserByEmail()
getUserKeywords()
updateSeveraOptInKeyword()

StudVladimir committed:
[add]: put getResourceAllocationsByUserOrAll() inside CreateSeveraApiService() fabric

StudVladimir committed:
[add]:  checkKeywordExists(),
fetchUserByEmail(),
getUserKeywords(),
updateSeveraOptInKeyword() inside CreateSeveraApiService() factory

Jdallos commented on src/services/severa-api-service.ts:
```diff
@@ -27,6 +27,11 @@ export interface SeveraApiService {
   getFilteredWorkHoursForUsers: (users: { guid: string }[], severaProjectId: string, startDate?: string, endDate?: string) => Promise<SeveraResponseWorkHours[]>;
   getWorkHoursForUser: (severaUserId: string, startDate?: string, endDate?: string) => Promise<SeveraResponseWorkHours[]>;
+  getResourceAllocationsByUserOrAll: (severaUserId: string | undefined, optInUsers: { guid: string }[]) => Promise<SeveraResponseResourceAllocation[]>;
+  checkKeywordExists: (keyword: string) => Promise<any>;
+  fetchUserByEmail: (email: string) => Promise<any>;
+  getUserKeywords: (userGuid: string) => Promise<any>;
+  updateSeveraOptInKeyword: (userGuid: string, isSeveraOptIn: string, isSeveraOptInKeywordGuid: string) => Promise<any>;
```
Lets avoid typing the return as any, we should be able to get the response type from the severa client, which it looks like is already set up in this project. There is also the use of SeveraResponse types e.g. SeveraResponseWorkhours which are custom types created by us, but again these could be replaced by the types generated from the client.

Considering this I think it would be reasonable to make sure the severa API spec is generated (ideally just the models rather than the Apis) and then this could also be committed as part of this PR.

Jdallos commented on src/services/severa-api-service.ts:
```diff
@@ -421,7 +426,213 @@ export const CreateSeveraApiService = (): SeveraApiService => {
+          method: "GET",
+          headers: {
+            Authorization: `Bearer ${await getSeveraAccessToken()}`,
+            Client_Id: process.env.SEVERA_DEMO_CLIENT_ID,
```
Could this also be a good PR to refactor the SEVERA_DEMO_CLIENT_ID env, we should add a SEVERA_CLIENT_ID env, both should be optional (and exclusive) so one must be present, both will not be.
Locally we will only be using the demo client but for the production we will add the production id (and not the demo id)

Jdallos replied:
This update should be done everywheere the demo client is currently used to make sure it will work when we switch to production environemnt

StudVladimir committed:
[fix]: strong 'return' typing,
[add]: util function getSeveraClientId() for managing production and develop env

Jdallos commented on src/services/severa-api-service.ts:
```diff
@@ -675,3 +678,10 @@ const getSeveraAccessToken = async (): Promise<string> => {
+
+// Utility function to get correct Severa client ID based on environment
+const getSeveraClientId = (): string => {
+  return process.env.NODE_ENV === "production"
```
I'm not certain that this will work as intended, as we have the staging environment which might also be classed as NODE_ENV === "production.
So we have 2 options, we handle this based on what envs are available in the pipeline, so we ensure only demo is in the staging pipeline and not demo id is in the production. This could work with a conditional selector for whichever one is truthy.

Doing at you are trying to do would be better, if we can confirm it is production in production but not in the staging env, perhaps the best way to check this is to add a log so we can see what the NODE ENV is when it's deployed to staging (vie develop merge).

Open to alternative suggestions

StudVladimir committed:
getSeveraClientId() use env.STAGE for defining Client_ID

StudVladimir committed:
[remove]: switch-case,  use if-else instead

Jdallos commented on src/services/severa-api-service.ts:
```diff
@@ -681,7 +681,9 @@ const getSeveraAccessToken = async (): Promise<string> => {
-  return process.env.NODE_ENV === "production"
-    ? process.env.SEVERA_CLIENT_ID
-    : process.env.SEVERA_DEMO_CLIENT_ID;
+  if (process.env.STAGE === "production") {
```
This is fine for now as we might end up changing this later to have a single env, and set it as appropriate depending on pipeline. 
Something to figure out during production review (or after)
